### ISSUE TITLE
Cursor interactivity is crippled when using tablet/stylus on Linux

### ISSUE THREAD

**Issue by brisvag** [napari/napari#5709]
## 🐛 Bug

When using a tablet+stylus to interact with the layers (other than pan-zoom), there is a huge input lag and a general "messiness" in the reaction of the canvas. 

What's odd is that not only pan-zoom works seamlessly, but using the mouse *in the same session* works perfectly fine. Here a small video showcasing the problem: the first shape is made with the mouse, while the second is with the stylus. As you can see, when moving the lifted stylus, the shape is updating *very slowly*, and sometimes jumps around. At the end of the second shape it's not even registering a bunch of clicks properly.

https://user-images.githubusercontent.com/23482191/231112393-6b645c81-f1dc-467c-832b-b18b4b5acc3f.mp4

Note that however the pan zoom at the end works perfectly responsively with both inputs.

The tablet also perfectly on other programs (like gimp).

---

I realize it's hard for others to debug something like this, but I'd appreciate some pointers on where to even begin :P

**Comment by jni** [comment]
I think @DragaDoncila has a wacom tablet? 😃 So maybe she can at least check whether she can reproduce?

I wonder whether the tablet produces many more events per second than the mouse (would make sense for a drawing tablet) and that is why napari is struggling?

**Comment by Czaki** [comment]
My first assumption is that the Wacom tablet generates more events than the mouse. And we may need to speed up event processing or find a way to query the queue and omit part of events. 

But this is the only an impression.

**Comment by jni** [comment]
haha jinx! 😂

**Comment by brisvag** [comment]
Hah! I didn't think of that... I'll see if I can check!

**Comment by brisvag** [comment]
I set up a simple even counter, and I'm not sure that's the problem... I tried to do as close as possible the same gesture with pen an mouse (including entering/exiting the canvas) and I end up with comparable number of events sent to the canvas (`v.window._qt_viewer.canvas.events`), maybe like 20% more with the stylus. That doesn't seem enough to warrant that lag :/

**Comment by Czaki** [comment]
Hard to debug without device :(

**Comment by brisvag** [comment]
Yes :( For now, I will try to use it on a different machine to see if the problem persists.

**Comment by brisvag** [comment]
Welp... On a different machine it works just fine. The broken one is a kubuntu 22.04, and the one that works is my regular Arch Linux machine. I don't know what to think now xD

**Comment by Czaki** [comment]
Maybe the tablet driver is somehow broken?

**Comment by brisvag** [comment]
Maybe... other option: my machine has a more powerful CPU and I'm no longer bottlenecked by it.

**Comment by brisvag** [comment]
Aha! I thinkl that was it. I tried being a bit more intense with the movements and use a layer that needs heavier calculations (polygonal chape, which calculates all the triangles for the filling) and I could reproduce on my machine.

**Comment by DragaDoncila** [comment]
> I think @DragaDoncila has a wacom tablet? 😃 So maybe she can at least check whether she can reproduce?

Sorry folks, I could not reproduce on any machine in the house - Mac, Ubuntu, big strong PC Windows and weak terrible laptop Windows, so I can't help. It sounds like it's been confirmed to be the mouse events though.

**Comment by brisvag** [comment]
I did see *similar* issues online with people using other software, and it seemed to be happening mostly with hi-res tablets (this one is 3840 x 2160 in 27"). Also, windows might handle things differently... could you try creating a big filled polygon shape and see if it gets slow once the number of faces increases?

**Comment by seankmartin** [comment]
I tried on Windows with a Wacom Intuos and drawing a big filled polygon with the pen and couldn't replicate the slowdown

**Comment by psobolewskiPhD** [comment]
@brisvag is this closed by #5710 ? Reading through, I think you were the only one affected.

**Comment by brisvag** [comment]
It's *reduced* by #5710 (now usable, but still not a seamless experience). I didn't want to throttle too much for now, but depending on the power of their computer and complexity fo their layer, this may show up again. Maybe we can leave open for discoverability?

Also, related to this: the problem is still there when itneracting with qt widgets instead of the canvas (for example, dragging the dims slider is basically impossible with the stylus... will likely be less of an issue with async loading).

**Comment by psobolewskiPhD** [comment]
I think there have been some changes in event handling over the past 2 years, e.g. things are done in a thread now.
@brisvag is this still an issue for you? it seems like you were the only one who was affected.

**Comment by brisvag** [comment]
I have no longer acess to that tablet unfortunately, so I cannot test this.

**Comment by psobolewskiPhD** [comment]
Oh! So you're not having the issue anymore! Resolved! ✅

**Comment by brisvag** [comment]
Solution: downgrade your hardware xD

**Comment by psobolewskiPhD** [comment]
@brisvag I think you mentioned having a tablet again? 😉

**Comment by brisvag** [comment]
So... good news with caveat, and straight up bad news.

Good news: no performance issue! Caveat: using a much more powerful machine, with a much cheaper tabled (so potentially much lower polling rate), which makes these good news pretty useless imo xD

Bad news: napari does not receive tabled events correctly *at all*, I suspect because of using multiple monitors and having mapped the tablet to one? The cursor position is nowhere near where the click actually happens. This does not happen in any other application, so it does appear to be a napari issue. Ouch.
